release-notes/rascal-0-41-x-release-notes.md

```diff
@@ -0,0 +1,606 @@
+* `util::Monitor::job` now throws a useful exception rather than `CallFailed` if the function or closure that is passed accidentally returns `void` (which is _not_ allowed).
+* A number of features were added to `lang::json::IO` which also impact `Content` servers and `util::WebServer` positively and `util::LanguageServer`. We can now print certain `data` types to string, just-in-time before serializing to JSON, and also parse that data back when it comes back from a web client (for example). Also it's possible to make all constructors explicit or just their data types as the first field of an object: `{ _type="Expression" }`. This comes in handy sometimes on the Typescript or Javascript side of things. On the way back, these fields are used for validation purposes.
+* Parse error reporting in `lang::json::IO` was radically improved, and origin locations along with them (they use the same internal administration of parsing positions).
+* `lang::json::IO` now supports mapping `null` to Rascal and back in different ways: it writes nothing at all (not even the key) for absent keyword parameters, and for `Maybe[void]::nothing()` it writes `null`. On the way back `null` becomes `Maybe[void]::nothing()` if a `Maybe[&T]` was expected, and keyword parameters are not bound if the value with a key was `null`. Also `null` elements in arrays are skipped. So if `null`s are expected in the input, use `Maybe[&T]` or keyword fields with defaults to model those on the Rascal side. 
+* `lang::json::IO` squeezes arbitrary precision `real`'s into JSON's floats. Integers are kept as-is, because JSON supports
+any integer and does not have overflow. Of course on the client side you will still have to deal with overflow when the number
+is bound to a field or variable.
+* `lang::json::IO`'s eight year old deprecated functions have finally been removed completely.
+* `Content` server now serves long (html or json) strings using a the internal concat/indent stream. This halfs the memory 
+consumption for large files, and still uses the fast internal streaming `Reader` on the implementations of `IString`. Also
+large files can now already be received and consumed by the client, in parallel, while the Rascal server is still sending. This has a positive effect on response times on the client side for web applications on the REPL with `Content`, using `util::Webserver` and the Language Service Protocol with `util::LanguageServer`.
+
+### Other Rascal Interpreter changes
+
+* Throwing and catching StackOverflow and OutOfMemory is possible again, due to not triggered OutOfMemory or StackOverflow during the handling of these exceptions.
+* The `visit` statement now _always_ memoizes amb clusters; this brings down the worst-case complexity of a visit with nested ambiguity to polynomial numbers (instead of the previous exponential amounts of nested combinations).
```

Comment:
```suggestion
* The `visit` statement now _always_ memoizes `amb` clusters; this brings down the worst-case complexity of a visit with nested ambiguity to polynomial numbers (instead of the previous exponential amounts of nested combinations).
```